# Telecom Churn Case Study

###### Problem Statement

*In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of <b>15-25% annual churn rate</b>. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.*

###### Task to Perform
To reduce customer churn, telecom companies need to predict which customers are at high risk of churn.

###### Understandings

There are two main models of payment in the telecom industry 
- postpaid (customers pay a monthly/annual bill after using the services)
- prepaid (customers pay/recharge with a certain amount in advance and then use the services)

***Postpaid***: when customers want to switch to another operator, they usually inform the existing operator to terminate the services, and you directly know that this is an instance of churn.

***Prepaid***: customers who want to switch to another network can simply stop using the services without any notice, and it is hard to know whether someone has actually churned or is simply not using the services temporarily.

#### Step1: Data Exploration and Preparation

In [1]:
# Importing the required library
import numpy as np
import pandas as pd

***Loading Data using pandas***

In [2]:
# Reading the Telecom dataset
telecom = pd.read_csv('./telecom_churn_data.csv')

# Displaying the top few reords
telecom.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6  ...  sachet_3g_9  fb_user_6  fb_user_7  \
0            9/30/2014  197.385  ...            0        1.0        1.0   
1            9/30/2014   34.047  ...            0        NaN        1.0   
2            9/30/2014  167.690  ...            0        NaN        NaN   
3            9/30/2014  221.338  ...            0        NaN        NaN   
4            9/30/2014  261.636  ...            0        0.0        NaN   

   fb_user_8  fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  
0        1.0        NaN   968        30.4         0.0      101.20        3.58  
1        1.0        NaN  1006         0.0         0.0        0.00        0.00  
2        NaN        1.0  1103         0.0         0.0        4.17        0.00  
3        NaN        NaN  2491         0.0         0.0        0.00        0.00  
4        NaN        NaN  1526         0.0         0.0        0.00        0.00  

[5 rows x 226 columns]

###### Basic Information

- *Analysing the Dataset about the datatype, columns etc*

In [3]:
telecom.shape

(99999, 226)

In [4]:
telecom.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


***Findings:***

- The dataset contains 99,999 rows and 226 columns
- Out of 226 columns 179 float, 35 int and 12 object datatype

### Exploratory Data Analysis

#### Step1: Data Cleaning

- First we will drop the columns that are not useful for our analysis like the columns that contains only one value

In [5]:
# Fetching the columns that contains same date in entire column
cols_no_unique = []

for col in telecom.columns:
    if telecom[col].nunique() == 1:
        cols_no_unique.append(col)

cols_no_unique

['circle_id',
 'loc_og_t2o_mou',
 'std_og_t2o_mou',
 'loc_ic_t2o_mou',
 'last_date_of_month_6',
 'last_date_of_month_7',
 'last_date_of_month_8',
 'last_date_of_month_9',
 'std_og_t2c_mou_6',
 'std_og_t2c_mou_7',
 'std_og_t2c_mou_8',
 'std_og_t2c_mou_9',
 'std_ic_t2o_mou_6',
 'std_ic_t2o_mou_7',
 'std_ic_t2o_mou_8',
 'std_ic_t2o_mou_9']

- Dropping the above columns as it has only single value that may not be useful for our analysis

In [6]:
# Dropping the columns
telecom = telecom.drop(cols_no_unique, axis=1)

### Treating Null Values

We will Start by looking at the percentage of null records in each column

In [7]:
# Percentage of null value analysis more than 0
(telecom.isna().mean() * 100)[(telecom.isna().mean() * 100) != 0]

onnet_mou_6          3.937039
onnet_mou_7          3.859039
onnet_mou_8          5.378054
onnet_mou_9          7.745077
offnet_mou_6         3.937039
                      ...    
night_pck_user_9    74.077741
fb_user_6           74.846748
fb_user_7           74.428744
fb_user_8           73.660737
fb_user_9           74.077741
Length: 152, dtype: float64

In [8]:
categorical_data = telecom.select_dtypes(include=['object'])
categorical_data.head()

date_of_last_rech_6 date_of_last_rech_7 date_of_last_rech_8  \
0           6/21/2014           7/16/2014            8/8/2014   
1           6/29/2014           7/31/2014           8/28/2014   
2           6/17/2014           7/24/2014           8/14/2014   
3           6/28/2014           7/31/2014           8/31/2014   
4           6/26/2014           7/28/2014            8/9/2014   

  date_of_last_rech_9 date_of_last_rech_data_6 date_of_last_rech_data_7  \
0           9/28/2014                6/21/2014                7/16/2014   
1           9/30/2014                      NaN                7/25/2014   
2           9/29/2014                      NaN                      NaN   
3           9/30/2014                      NaN                      NaN   
4           9/28/2014                 6/4/2014                      NaN   

  date_of_last_rech_data_8 date_of_last_rech_data_9  
0                 8/8/2014                      NaN  
1                8/10/2014                      NaN  
2                      NaN                9/17/2014  
3                      NaN                      NaN  
4                      NaN                      NaN